### DISCUSSION TITLE
Code coverage of a pybind11 library

### DISCUSSION THREAD

**Discussion by hman523** [pybind/pybind11#4141] [Q&A]
Hi all! I am trying to determine if I am testing all of the functions within a python module made using pybind.  I have tried using coverage.py but it only showed the functions written 100% in python. I also tried gcov but it did not really show anything meaningful.  I would like to just see what functions from the module are called when running pytest. Has anyone done this before?

**Comment by jpreiss** [comment]
Did you ever find a good solution?

**Comment by ianthomas23** [comment]
`contourpy` generates coverage information for both Python and pybind11-wrapped C++ in this workflow: https://github.com/contourpy/contourpy/blob/main/.github/workflows/test.yml (see the "Test debug with coverage" run). Log file for the latest run is at https://github.com/contourpy/contourpy/actions/runs/6220269820/job/16879953676 and it uploads to https://app.codecov.io/gh/contourpy/contourpy.

Essentially you need to

- Pass appropriate compile and link flags
- Use appropriate flags when running `pytest`
- Extract and combine the Python and C++ coverage information

If you want to do this locally rather than upload to codecov then you need to do something like
```
cat coverage.lcov coverage.cpp > coverage.combined
lcov --list coverage.combined
```
to get textual output or
```
genhtml coverage.combined --output-directory out
```
to generate HTML output.

Note this is using `meson-python` as the build backend. If you are using a different build backend you will no doubt need different compile/link flags.

**Comment by jpreiss** [comment]
@ianthomas23 thank you for the suggestions!!
